docs/docs/specs/v0/composer-2.md

```diff
@@ -0,0 +1,389 @@
+
+## Example Structure
+
+Below is a diagram illustrating how:
+
+- a full application is built up to handle a `Stack` of `Composer` and `Image`
+- `Surfaces` are the only element of abstraction necessary to both compose the entire application chrome and implement the abstract portion of `Stack`.
+
+![Composer Structural Diagram](./composer.drawio.svg)
+
+## Stacks
+
+- Stacks are just components which rely on the same `<Surface />` concept to delegate the concrete rendering of stack elements to components provided by plugins.
+
+## Routing
+
+- Routing (the main router and top level routes) are declared outside (around) the `<App />` component and allow the `App` to pass route information as state in the total `AppState` context for all descendants.
+- Nested components inside `<App />` are equally free to use `<Route />` elements if necessary as well.
+- Nested components are encouraged to perform navigation by submitting `Actions` as opposed to calling `navigate()` directly, although situations like `<Button to="/some/route"/>` should also be supported.
+
+## Application State and Actions
+
+- The App provides a global `AppState` (in the `AppContext`) to all nested components which is composed of some concrete and other dynamic sections to be filled by plugins
+- State is updated via a `dispatch(...actions)` API in the `AppContext` which enables Plugins to subscribe to each other's actions and update state accordingly
+- Plugins (and their provided Components) are free to extend/add more fields to the `AppState` as necessary, when it is desirable to share certain state application-wide.
+- State can be composed of `@preact/signals` (or similar reactive state solution) and can contain any other objects including `ECHO` objects.
+- When state is:
+  - local, ephemeral, per-session: it is managed by signals (and backed by local or session storage as appropriate). i.e. the selection state in a List or Tree is ephemeral and not shared between windows or devices, but may be shared among components and plugins in the application.
+  - peer-to-peer, permanent: it is managed by ECHO objects.
+- `AppState` includes current route and params in convenient form.
+- Actions are handled by functions that resemble `reducers` who should return new state reflecting the action committing, and possibly a list of side-effects which may result in more actions or state updates at a later time. (i.e. performing writes to echo objects in a batch is a side-effect, invoking a github API is a side-effect, etc. whereas the indicator that the app is loading something is a purely new state returned from the reducer)
+
+## Plugins, Components, Actions, and other `provides`
+
+- The application framework consumes a list of plugins (statically compiled-in for now).
+- Plugins can provide Components in response to a specific data context: i.e `plugin.provides.component(data: any): React.FC`
```

Comment:
(perhaps out of scope) do you think it would be worthwhile for the type of `data` be `unknown` and have the component validate the data shape before consuming it?

Reply:
yes exactly I expect components are likely validating the shape and returning null or nothing when they feel the data isn't applicable, we'll likely employ some patterns for validation such as zod for example

Reply:
So wouldn't `unknown` be more accurate and type-safe in this case then? Of course, wouldn't stop anyone else from using `any` or not even using TS but at least points to a suggested validation path